### PR TITLE
RCB-431_API-Builder

### PR DESCRIPTION
**Pull request by hgaudet**

### CODE REVIEW

File: api/src/main/java/com/basistech/rosette/api/RosetteAPI.java [hunk 1]
```diff
@@ -71,10 +73,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.io.Closeable;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.OutputStream;
+import java.io.*;
```
Comment (guruxu) [inline api/src/main/java/com/basistech/rosette/api/RosetteAPI.java]: We do not roll up imports as a convention

File: api/src/main/java/com/basistech/rosette/api/RosetteAPI.java [hunk 2]
```diff
@@ -166,6 +172,42 @@ public RosetteAPI(String key, String alternateUrl) throws IOException, RosetteAP
...
+     * @param genre          genre (ex "social-media")
+     * @param options        options for the request
+     * @throws IOException          General IO exception
+     * @throws RosetteAPIException  Problem with the API request
+     */
+    private RosetteAPI(String key, String alternateUrl, int failureRetries,
+                       LanguageCode language, String genre, Options options,
+                       CloseableHttpClient httpClient)
+                        throws IOException, RosetteAPIException {
+        Objects.requireNonNull(alternateUrl, "alternateUrl cannot be null");
```
Comment (guruxu) [inline api/src/main/java/com/basistech/rosette/api/RosetteAPI.java]: "alternate" is no longer appropriate in this context. It really is _the_url, whether api.rosette.com or not.

File: api/src/main/java/com/basistech/rosette/api/RosetteAPI.java [hunk 3]
```diff
@@ -166,6 +172,42 @@ public RosetteAPI(String key, String alternateUrl) throws IOException, RosetteAP
...
+        this.key = key;
+        this.language = language;
+        this.genre = genre;
+        this.options = options;
+        urlBase = alternateUrl;
+        if (urlBase.endsWith("/")) {
+            urlBase = urlBase.substring(0, urlBase.length() - 1);
```
Comment (guruxu) [inline api/src/main/java/com/basistech/rosette/api/RosetteAPI.java]: would `urlBase.trim().replaceAll("/+$", "")` be better if there are accidental trailing spaces and multiple slashes?

File: api/src/main/java/com/basistech/rosette/api/RosetteAPI.java [hunk 4]
```diff
@@ -339,16 +336,33 @@ public PingResponse ping() throws IOException, RosetteAPIException {
...
+     * @param request request
      * @return response
      * @throws RosetteAPIException Rosette specific exception
      * @throws IOException         General IO exception
+     *
+     * @deprecated replaced by {@link #getNameSimilarity(Name, Name) getNameSimilarity}
      */
+    @Deprecated
     public NameSimilarityResponse getNameSimilarity(NameSimilarityRequest request) throws RosetteAPIException, IOException {
         return sendPostRequest(request, urlBase + NAME_SIMILARITY_SERVICE_PATH, NameSimilarityResponse.class);
```
Comment (guruxu) [inline api/src/main/java/com/basistech/rosette/api/RosetteAPI.java]: Should just call the new method? `getNameSimilarity(request.getName1(), request.getName2())`

File: api/src/main/java/com/basistech/rosette/api/RosetteAPI.java [hunk 5]
```diff
@@ -1800,4 +1815,112 @@ public void close() throws IOException {
...
+         * Construct the request object.
+         * @return the request object.
+         */
+        public abstract RosetteAPI build() throws IOException, RosetteAPIException;
+    }
+
+    /**
+     * Fluent builder class for {@link DocumentRequest} objects.
+     */
+    public static class Builder<O extends Options> extends BaseBuilder<O, Builder<O>> {
```
Comment (guruxu) [inline api/src/main/java/com/basistech/rosette/api/RosetteAPI.java]: What's the reason to have a separate base and this?

**Comment by hgaudet** [comment]
In EntitiesExample I left a commented out import (EntitiesOptions) because users will need that import if they decide to set linking themselves

File: api/src/main/java/com/basistech/rosette/api/RosetteAPI.java [hunk 6]
```diff
@@ -166,6 +175,39 @@ public RosetteAPI(String key, String alternateUrl) throws IOException, RosetteAP
...
+     * @param genre          genre (ex "social-media")
+     * @param options        options for the request
+     * @throws IOException          General IO exception
+     * @throws RosetteAPIException  Problem with the API request
+     */
+    private RosetteAPI(String key, String url, int failureRetries,
+                       LanguageCode language, String genre, Options options,
+                       CloseableHttpClient httpClient)
+                        throws IOException, RosetteAPIException {
+        Objects.requireNonNull(url, "alternateUrl cannot be null");
```
Comment (guruxu) [inline api/src/main/java/com/basistech/rosette/api/RosetteAPI.java]: `url`?

File: api/src/main/java/com/basistech/rosette/api/RosetteAPI.java [hunk 7]
```diff
@@ -166,6 +175,39 @@ public RosetteAPI(String key, String alternateUrl) throws IOException, RosetteAP
...
+        urlBase = url.trim().replaceAll("/+$", "");
+        this.failureRetries = failureRetries;
+        mapper = ApiModelMixinModule.setupObjectMapper(new ObjectMapper());
+        customHeaders = new ArrayList<>();
+        if (httpClient == null) {
+            initHttpClient();
+        } else {
+            this.httpClient = httpClient;
+        }
+        checkVersionCompatibility();
```
Comment (guruxu) [inline api/src/main/java/com/basistech/rosette/api/RosetteAPI.java]: Do we still need to do this?

File: examples/src/main/java/com/basistech/rosette/examples/EntitiesExample.java [hunk 8]
```diff
@@ -35,16 +36,19 @@ public static void main(String[] args) {
 
     private void run() throws IOException, RosetteAPIException {
         String entitiesTextData = "Bill Murray will appear in new Ghostbusters film: Dr. Peter Venkman was spotted filming a cameo in Boston this… http://dlvr.it/BnsFfS";
-
-        RosetteAPI rosetteApi = new RosetteAPI(getApiKeyFromSystemProperty(), getAltUrlFromSystemProperty());
-        rosetteApi.setOptions(null);
-        rosetteApi.setGenre("social-media");
+        RosetteAPI rosetteApi = new RosetteAPI.Builder()
+                                .apiKey(getApiKeyFromSystemProperty())
+                                .alternateUrl(getAltUrlFromSystemProperty())
```
Comment (guruxu) [inline examples/src/main/java/com/basistech/rosette/examples/EntitiesExample.java:40]: what if this is null?

File: api/src/main/java/com/basistech/rosette/api/RosetteAPI.java (hunk 7 above)
Comment (hgaudet) [inline api/src/main/java/com/basistech/rosette/api/RosetteAPI.java]: Which section are you referring to?

File: examples/src/main/java/com/basistech/rosette/examples/EntitiesExample.java (hunk 8 above)
Comment (hgaudet) [inline examples/src/main/java/com/basistech/rosette/examples/EntitiesExample.java:40]: If api key is null, a rosette exception is thrown saying authentication is needed, I've added a check for a null alternateUrl

File: api/src/main/java/com/basistech/rosette/api/RosetteAPI.java (hunk 7 above)
Comment (guruxu) [inline api/src/main/java/com/basistech/rosette/api/RosetteAPI.java]: I thought we no longer check version compatibility

File: api/src/main/java/com/basistech/rosette/api/RosetteAPI.java [hunk 9]
```diff
@@ -166,6 +175,38 @@ public RosetteAPI(String key, String alternateUrl) throws IOException, RosetteAP
...
+     * @param failureRetries Number of times to retry in case of failure; default 1
+     * @param language       source language
+     * @param genre          genre (ex "social-media")
+     * @param options        options for the request
+     * @throws IOException          General IO exception
+     * @throws RosetteAPIException  Problem with the API request
+     */
+    private RosetteAPI(String key, String urlToCall, int failureRetries,
+                       LanguageCode language, String genre, Options options,
+                       CloseableHttpClient httpClient)
```
Comment (guruxu) [inline api/src/main/java/com/basistech/rosette/api/RosetteAPI.java]: Just `HttpClient`, it's up to the caller to provide the actual impl.

File: api/src/main/java/com/basistech/rosette/api/RosetteAPI.java [hunk 10]
```diff
@@ -1800,4 +1814,101 @@ public void close() throws IOException {
...
+     * Builder class for the RosetteAPI object.
+     */
+    public static class Builder {
+        protected LanguageCode language;
+        protected String genre;
+        protected Options options;
+        protected String key;
+        protected String urlBase = DEFAULT_URL_BASE;
+        protected int failureRetries = 1;
+        protected CloseableHttpClient httpClient;
```
Comment (guruxu) [inline api/src/main/java/com/basistech/rosette/api/RosetteAPI.java]: `HttpClient`

File: examples/src/main/java/com/basistech/rosette/examples/EntitiesExample.java (hunk 8 above)
Comment (guruxu) [inline examples/src/main/java/com/basistech/rosette/examples/EntitiesExample.java:40]: You can also do it in the builder.

**Pull request merged**